test/metabase/upload_test.clj

```diff
@@ -1230,10 +1231,18 @@
   (try
     (mt/with-temporary-setting-values [uploads-enabled uploads-enabled]
       (mt/with-current-user user-id
-        (let [table-id (or table-id (:id (create-upload-table!)))]
-          (t2/update! :model/Table table-id {:is_upload is-upload})
-          (append-csv! {:table-id table-id
-                        :file     file}))))
+        (mt/with-model-cleanup [:model/Table]
+          (let [new-table (when (nil? table-id)
+                            (create-upload-table!))
+                table-id (or table-id (:id new-table))]
+            (t2/update! :model/Table table-id {:is_upload is-upload})
+            (try (append-csv! {:table-id table-id
+                               :file     file})
+                 (finally
+                   (when new-table
```

Comment:
Do not be fooled by the length of this comment, it's just a nit.

On an initial reading this was not clear to me why the clean up code only existed for one case, and started writing a comment here. The penny dropped before I posted, but I think it's a natural question mark for a reader, and short comment about how the cases are handled differently would be nice. Perhaps something like "mt/with-model-cleanup retracts inserted rows, but we must manually retract new tables". 

In general it would be nice to use the `with` macro pattern consistently, which would also group all the clean-up together. It would be nice to cover file deletion this way too.
  
This is very much gilding the lily, given the narrow scope of this helper, but since I already started writing a comment, I think it's worth sharing the guidance. This is yet another example of the "single level of abstraction" principle I'm so fond of - `finally` is a lower level of abstraction than `with`, so they don't belong in the same function.

If we want to avoid proliferating macros, the idiom in JVM languages without macros is the "try with resources" pattern, i.e. `with-open`, combined with wrappers for files, tables etc that implement`AutoCloseable`. We could probably use a protocol for values with `:close-fn` metadata, since I think that macro just uses duck typing. What's nice about that pattern is we'd get leverage through the whole code base with just a wrapper per type, and we can indent just one no matter how many resources we are cleaning up.

Reply:
Thanks for the detailed explanation! 

The reason why the `with-upload-table!` macro isn't being used here is because the table needs to be conditionally created depending on whether `table-id` is supplied as an argument. I can't think of an elegant way to get around that right now, but you might have an idea.

I'm going to make a mental note to come back to this technique in the future anyway. I haven't seen any implementations of AutoCloseable at metabase before.